addons/website/tests/test_lang_url.py

```diff
@@ -155,3 +156,56 @@ def assertUrlRedirect(url, expected_url, msg="", code=301):
         # website.page
         assertUrlRedirect('/fr/page_1/', '/fr/page_1', "Check for website.page with language in URL.")
         assertUrlRedirect('/fr/page_1/?a=b', '/fr/page_1?a=b', "Check for website.page with language in URL + URL params.")
+
+@tagged('-at_install', 'post_install')
+class TestTranslateUrl(TestLangUrl):
+    def test_translate_url(self):
+        base = "http://%s:%s" % (common.HOST, config['http_port'])
+        view_test_translate_url = self.env['ir.ui.view'].create({
+            'name': 'Base',
+            'type': 'qweb',
+            'arch': '<div>English page</div>',
+            'key': 'test.view_test_translate_url',
+        })
+        page = self.env['website.page'].create({
+            'view_id': view_test_translate_url.id,
+            'url': '/page-en',
+            'is_published': True,
+        })
+        page.with_context(lang='fr_FR').url = '/page-fr'
+        homepage_domain = [('url', '=', '/')] + self.website.website_domain()
+        homepage_specific = self.env['website.page'].search(homepage_domain, order='website_id asc', limit=1)
+        homepage_specific.with_context(lang='fr_FR').url = '/accueil'
+
+        # Trying to access the french url of a page (without the lang in the
+        # url) if the website language is in english should redirect to the
+        # english url of this page.
+        r = self.url_open('/page-fr')
+        self.assertEqual(r.history[0].status_code, 303)
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.url, base + '/page-en')
+
+        # Go to the homepage.
+        r = self.url_open('/')
+        # From the homepage, changing the language of the website should
+        # redirect to the french url of the specific homepage.
+        r = self.url_open('/fr')
+        # self.assertEqual(r.history[0].status_code, 303)
+        self.assertEqual(r.status_code, 200)
+        self.assertEqual(r.url, base + '/fr/accueil')
```

Comment:
Note: This test does not work for the moment. Wait for [this task](https://www.odoo.com/web#id=3476840&menu_id=6444&cids=1&action=5050&model=project.task&view_type=form) to be finished for it to work.